File: cli/src/commands/sync-test-data.ts
import { getRootedPath } from '@/utils/file_utils.js';
import fs from 'fs';

import { CommandBase } from './command_base.js';

export class SyncTestData extends CommandBase {
  constructor() {
    super(import.meta.url);

    this.description('Sync test data');
  }

  override async run() {
    await downloadJsonAndProtoFor(process.env.test_api_key!, 'eval_proj_dcs');

    await downloadJsonAndProtoFor(process.env.PERF_SDK_KEY!, 'perf_proj_dcs');

    await downloadJsonAndProtoFor(
      process.env.statsig_demo_server_key!,
      'demo_proj_dcs',
    );
  }
}

async function downloadJsonAndProtoFor(sdkKey: string, name: string) {
  downloadFromUrlToFile(
    `https://staging.statsigapi.net/v2/download_config_specs/${sdkKey}.json`,
    `${name}.json`,
    'application/json; charset=utf-8',
  );
  downloadFromUrlToFile(
    `https://staging.statsigapi.net/v2/download_config_specs/${sdkKey}.json?supports_proto=1`,
    `${name}.pb.br`,
    'application/octet-stream',
  );
}

async function downloadFromUrlToFile(
  url: string,
  filename: string,
  contentType: string,
) {
  const response = await fetch(url);

  if (response.headers.get('Content-Type') !== contentType) {
    throw new Error(
      `Expected content type ${contentType} but got ${response.headers.get(
        'Content-Type',
      )}`,
    );
  }

  const data =
    contentType === 'application/octet-stream'
      ? await response.arrayBuffer()
      : await response.text();

  writeToDataDir(filename, data);
}

function writeToDataDir(filename: string, data: string | ArrayBuffer) {
  const path = getRootedPath(`statsig-rust/tests/data/${filename}`);
  fs.writeFileSync(
    path,
    data instanceof ArrayBuffer ? Buffer.from(data) : data,
  );
}

File: statsig-rust/tests/bin/sync_test_data.rs
#[tokio::main]
async fn main() {
    // CompanyID: 5NprLGRxV3W28hreG51Z7n
    let eval_proj_key = std::env::var("test_api_key").unwrap();
    download_json_and_proto_for(&eval_proj_key, "eval_proj_dcs").await;

    // CompanyID: 3ncq00rOXvRdibzZTVUDgl
    let perf_sdk_key = std::env::var("PERF_SDK_KEY").unwrap();
    download_json_and_proto_for(&perf_sdk_key, "perf_proj_dcs").await;

    // CompanyID: 2etz0PtUkhGsJfH0mR6whu
    let demo_sdk_key = std::env::var("statsig_demo_server_key").unwrap();
    download_json_and_proto_for(&demo_sdk_key, "demo_proj_dcs").await;
}

async fn download_json_and_proto_for(sdk_key: &str, name: &str) {
    download_from_url_to_file(
        format!("https://staging.statsigapi.net/v2/download_config_specs/{sdk_key}.json"),
        format!("{name}.json"),
    )
    .await;

    download_from_url_to_file(
        format!("https://staging.statsigapi.net/v2/download_config_specs/{sdk_key}.json?supports_proto=1"),
        format!("{name}.pb.br"),
    )
    .await;
}

async fn download_from_url_to_file(url: String, filename: String) {
    let response = reqwest::get(&url).await.unwrap();
    let body = response.text().await.unwrap();
    write_to_data_dir(&filename, &body);
}

fn write_to_data_dir(filename: &str, body: &str) {
    let mut path = std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("tests/data");
    path.push(filename);
    std::fs::write(path, body).unwrap();
}
